### DISCUSSION TITLE
how to setMaxSpeed..?

### DISCUSSION THREAD

**Discussion by andymid3D** [gin66/FastAccelStepper#287] [Q&A] [answered]
Hi, 
reading thru https://github.com/gin66/FastAccelStepper/blob/master/extras/doc/FastAccelStepper_API.md, setting speed "Invalid is faster than MaxSpeed" which I understand is related to the max hardware limits, but I would like to implement a safety so that the user cannot setSpeed higher than my chosen maxSpeed, which is far below the limit of the ESP32 I am using, but would make the machine I am making (even more) dangerous than it already is (using a nema 34 160mm, with 6A driver, and a 20:1 reduction to inject plastic with a plunger, too fast moves could conceivably apply too much force to the drive chain, or if hitting (inductive) endstops, not give enough time to actually stop the inertia of the system...

How can I add a setMaxSpeed in the setup()..? I understand that later I would not even need a check as the library, for the reading above indicated, would not allow a setSpeed above the MaxSpeed..?
Otherwise I can set my own MaxSpeed, and implement a compare function always, but wondered if this was possible already within this library 

Which by the way I am very happy to be using after trying at least 3 others, this is by far the fastest and smoothest movement (although the use of interrupts and a independent encoder that also used interrupts has forced me to abandon using a Nano, to use now a ESP32, this has been also a beneficial learning process)

thanks!
Andy
:-)

**Comment by gin66** [comment]
How about `setAbsoluteSpeedLimit(uint16_t max_speed_in_ticks);` ?

**Reply by andymid3D** [reply]
Thanks for the quick reply! 
as this command was expressly marked as untested, wasn’t sure that it would be safe, but I can test  ;-)

to be clear, I am using all speed values in hz, so my max of 5000 step/s = 3200 ticks..? 16.000.000 tick/s  /  5000steps/s = 3200 ticks/ step…? 

Thanks! 
Andy
:-)

**Reply by gin66** [reply, marked as answer]
That is the right calculation. 

The warning is for (accidental) misuse. This call allows to effectively set a maximum speed way above, what the uC can do (with or without load).There is no check to go too high. The API user is warned…

**Reply by andymid3D** [reply]
Perfect! so setting maxSpeed way below what the uC can do is not an issue here, cool!

**Reply by gin66** [reply]
As the intention of the call was to exceed the speed limit set by FastAccelStepper, the use case of very low speed has due to the `uint16`_t a lower limitation: ticks=65535 translates into 244steps/s for 16MHz tick rate.

**Reply by andymid3D** [reply]
noted, there is a max ticks value (65535) that may be used, but as my desired MaxSpeed is above (and ticks below?) this lower limitation as well (3200 ticks = 5000steps/s), then in theory should not be an issue either..? useful info to have been informed about thou ;-)
Thanks again!

**Comment by andymid3D** [comment]
I got an email about a new post here...
I was about to comment* on the message just posted "to find your solution open this tutor site : https:// www. mandevilletrailheadmarket .com/" <- link broken with spaces so no-one clicks!
but it has already been deleted.. 👏👏👏
the user who posted is suspicious... 
* BEWARE! phishing has reached Github! 

![Screenshot 2024-11-01 at 10 08 58](https://github.com/user-attachments/assets/7fc6eea3-764d-4bbd-beb8-da4cf422ae49)

**Reply by gin66** [reply]
yes. I have deleted it, blocked  and reported the user

**Comment by andymid3D** [comment]
@gin66 FYI, tried installing ESP32 core 3.1.0 RC2, but example sketch PrintPosition now fails to compile.. have already installed your latest library 0.31.2, no difference.. 
gives Compilation error: exit status 126, and above says:
bash: /Users/andy/Library/Arduino15/packages/esp32/tools/esptool_py/4.8.1/esptool: Bad CPU type in executable

took board version back to RC1 and now compiles as before, so will stay with that FTM ;-)

**Reply by gin66** [reply]
`Bad cpu type` does not hint to a problem with FastAccelStepper

**Comment by andymid3D** [comment]
and whilst am here, would this code be correct:

  engine.init(1); // want to dedicate 1 core to motor driver only, to avoid any type of delays from many digitalReads that have to do (although not every loop)
  stepper = engine.stepperConnectToPin(stepPinStepper, 1); // want to use RMT as want to use independent encoder using PCNT and your docs say RMT is better in the case of choice...

**Reply by gin66** [reply]
looks ok
